app/services/topbar_sweeper.rb

```diff
@@ -0,0 +1,22 @@
+class TopbarSweeper
+  include Singleton
+  attr_accessor :controller
+
+  def expire_cache(user = User.current)
+    controller.expire_fragment(self.class.fragment_name(user.id)) if controller.present? && user.present?
+  end
+
+  class << self
+    delegate :expire_cache, to: :instance
+
+    def fragment_name(id = User.current.id)
+      "tabs_and_title_records-#{id}"
+    end
+
+    def expire_cache_all_users
+      User.unscoped.pluck(:id).each do |id|
+        Rails.cache.delete("views/#{fragment_name(id)}")
+      end
+    end
+  end
+end
```

Comment:
Yes, I know we are not thread safe, but that does not mean we should avoid cleaner design. But it looks like I missed the fact we are observing models where controllers are not accessible. Then it makes sense, let me do one more read if I find something better.